### PR TITLE
Convert bucket aggs docs to runtime fields

### PR DESCRIPTION
**Pull request by nik9000**
This replaces the `script` docs for bucket aggregations with runtime
fields. We expect runtime fields to be nicer to work with because you
can also fetch them or filter on them. We expect them to be faster
because their don't need this sort of `instanceof` tree:
https://github.com/elastic/elasticsearch/blob/a92a647b9f17d1bddf5c707490a19482c273eda3/server/src/main/java/org/elasticsearch/search/aggregations/support/values/ScriptDoubleValues.java#L42

Relates to #69291

### CODE REVIEW

**Review by imotov — APPROVED** [review]
Nice!

File: docs/reference/aggregations.asciidoc [hunk 1]
```diff
@@ -330,79 +330,65 @@ the aggregated field.
 [[use-scripts-in-an-agg]]
 === Use scripts in an aggregation
 
-Some aggregations support <<modules-scripting,scripts>>. You can
-use a `script` to extract or generate values for the aggregation:
+When a field doesn't exactly match the aggregation you need to do you
+should aggregate on a <<runtime,runtime field>>:
```
Comment (lockewritesdocs) [inline docs/reference/aggregations.asciidoc:334]:
```suggestion
When a field doesn't exactly match the aggregation you need, you
should aggregate on a <<runtime,runtime field>>:
```

File: docs/reference/aggregations.asciidoc [hunk 2]
```diff
@@ -330,79 +330,65 @@ the aggregated field.
...
-* `date`
-* `double`, used for all floating-point numbers
-* `long`, used for all integers
-* `ip`
-* `string`
+Scripts will calculate the values on the fly, adding a little overhead
+to the execution of the aggretion. Some aggregations like
+<<search-aggregations-bucket-terms-aggregation,`terms`>> and
+<<search-aggregations-bucket-filters-aggregation,`filters`>> have to
+give up some optimizations.
```
Comment (lockewritesdocs) [inline docs/reference/aggregations.asciidoc:389]:
```suggestion
Scripts calculate field values dynamically, which adds overhead
to the aggregation. Certain aggregations like
<<search-aggregations-bucket-terms-aggregation,`terms`>> and
<<search-aggregations-bucket-filters-aggregation,`filters`>> are
less optimized when using scripts.
```

Comment (lockewritesdocs) [reply docs/reference/aggregations.asciidoc:389]: @nik9000, let me know if I totally misinterpreted your meaning here.

File: docs/reference/aggregations/bucket/composite-aggregation.asciidoc [hunk 3]
```diff
@@ -142,33 +141,66 @@ GET /_search
 }
 --------------------------------------------------
 
-Like the `terms` aggregation it is also possible to use a script to create the values for the composite buckets:
+Like the `terms` aggregation it is also possible to use a
+<<runtime,runtime field>> to create the values for the composite buckets:
```
Comment (lockewritesdocs) [inline docs/reference/aggregations/bucket/composite-aggregation.asciidoc:145]:
```suggestion
Like the `terms` aggregation, it's possible to use a
<<runtime,runtime field>> to create values for the composite buckets:
```

File: docs/reference/aggregations/bucket/composite-aggregation.asciidoc [hunk 4]
```diff
@@ -197,25 +229,35 @@ GET /_search
 }
 --------------------------------------------------
 
-The values are built from a numeric field or a script that return numerical values:
+Like the `histogram` aggregation it is also possible to use a
+<<runtime,runtime field>> to create the values for the composite buckets:
```
Comment (lockewritesdocs) [inline docs/reference/aggregations/bucket/composite-aggregation.asciidoc:233]:
```suggestion
Like the `histogram` aggregation it's possible to use a
<<runtime,runtime field>> to create values for the composite buckets:
```

File: docs/reference/aggregations/bucket/datehistogram-aggregation.asciidoc [hunk 5]
```diff
@@ -621,9 +621,75 @@ Response:
 [[date-histogram-scripts]]
 ==== Scripts
 
-As with the normal <<search-aggregations-bucket-histogram-aggregation,histogram>>,
-both document-level scripts and
-value-level scripts are supported. You can control the order of the returned
+Use a <<runtime,runtime field>> if the data in your documents doesn't
+exactly match what you'd like to aggregate. If, for example, revenue
+for promoted sales should be recognized a day after their sale date:
```
Comment (lockewritesdocs) [inline docs/reference/aggregations/bucket/datehistogram-aggregation.asciidoc:626]:
```suggestion
If the data in your documents doesn't exactly match what you'd like to aggregate,
use a <<runtime,runtime field>> . For example, if the revenue
for promoted sales should be recognized a day after the sale date:
```

File: docs/reference/aggregations/bucket/datehistogram-aggregation.asciidoc [hunk 6]
```diff
@@ -673,51 +739,52 @@ the `order` setting. This setting supports the same `order` functionality as
 [[date-histogram-aggregate-scripts]]
 ===== Using a script to aggregate by day of the week
 
-When you need to aggregate the results by day of the week, use a script that
-returns the day of the week:
+When you need to aggregate the results by day of the week run a `terms`
+aggregation on a <<runtime,runtime field>> that returns the day of the week.
```
Comment (lockewritesdocs) [inline docs/reference/aggregations/bucket/datehistogram-aggregation.asciidoc:743]:
```suggestion
When you need to aggregate the results by day of the week, run a `terms`
aggregation on a <<runtime,runtime field>> that returns the day of the week:
```

File: docs/reference/aggregations/bucket/range-aggregation.asciidoc [hunk 7]
```diff
@@ -172,22 +172,31 @@ Response:
 
 ==== Script
 
-Range aggregation accepts a `script` parameter. This parameter allows to defined an inline `script` that
-will be executed during aggregation execution.
+Use a <<runtime,runtime field>> if the data in your documents doesn't
+exactly match what you'd like to aggregate. If, for example, you need to
+apply a particular currency conversion rate:
```
Comment (lockewritesdocs) [inline docs/reference/aggregations/bucket/range-aggregation.asciidoc:177]:
```suggestion
If the data in your documents doesn't exactly match what you'd like to aggregate,
use a <<runtime,runtime field>>. For example, if you need to
apply a particular currency conversion rate:
```

**Review by lockewritesdocs — APPROVED** [review]
Awesome @nik9000! I left a few nits and some suggested editorial changes.

File: docs/reference/aggregations.asciidoc (hunk 2 above)
Comment (nik9000) [reply docs/reference/aggregations.asciidoc]: Yeah, this is right. I think I meant something a little more nuanced with "a little" than comes across on rereading. I'll merge you suggestion and then iterate on it locally for a bit.

**Review by jrodewig — APPROVED** [review]
👍  This looks great to me. Thanks for converting these, @nik9000.

**Comment by nik9000** [comment]
Thanks for the reviews! I'm going to merge these in the morning and grab some more. I figure it'd be nice to get these done, at least in aggs and search land.

File: docs/reference/aggregations.asciidoc [hunk 8]
```diff
@@ -330,79 +330,66 @@ the aggregated field.
...
-* `double`, used for all floating-point numbers
-* `long`, used for all integers
-* `ip`
-* `string`
+Scripts calculate field values dynamically, which adds a little
+overhead to the aggregation. In addition to the time spent calculating,
+some aggregations like <<search-aggregations-bucket-terms-aggregation,`terms`>>
+and <<search-aggregations-bucket-filters-aggregation,`filters`>> can not use
+some of their optimizations with runtime fields. So the total performance cost
+for using a runtime field varies from aggregation to aggregation.
```
Comment (lockewritesdocs) [inline docs/reference/aggregations.asciidoc:390]:
```suggestion
Scripts calculate field values dynamically, which adds a little
overhead to the aggregation. In addition to the time spent calculating,
some aggregations like <<search-aggregations-bucket-terms-aggregation,`terms`>>
and <<search-aggregations-bucket-filters-aggregation,`filters`>> can't use
some of their optimizations with runtime fields. In total, performance costs
for using a runtime field varies from aggregation to aggregation.
```

Comment (lockewritesdocs) [reply docs/reference/aggregations.asciidoc:390]: I'd suggest "cannot" or "can't" (probably "can't"), and the other suggestion is just trying to lead into that last sentence a bit more.

Comment (nik9000) [reply docs/reference/aggregations.asciidoc:390]: :+1:

**Pull request merged**